### PR TITLE
Support LogDNA archiving

### PR DESCRIPTION
**Pull request by shemau**
### Description

A new feature to support archiving of LogDNA and Activity Tracker events to IBM Cloud Object Storage.

- Add new properties to allow an existing COS bucket to be passed in for the LogDNA instance and/or the Activity tracker instance
- Include a new example `observability_archive` to cover usage for a LogDNA instance

#752

#### No release required

- [ ] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other changes that don't affect Terraform code

#### Release required

- [ ] Bug fix (patch release (`x.x.X`): Change that fixes an issue and is compatible with earlier versions)
- [x] New feature (minor release (`x.X.x`): Change that adds functionality and is compatible with earlier versions)
- [ ] Breaking change (major release (`X.x.x`): Change that is likely incompatible with previous versions)

##### Release notes content

When upgrading to this release new providers are required

    provider "logdna" {
      alias      = "at"
      servicekey = module.test_observability_instance_creation.activity_tracker_resource_key != null ?  module.test_observability_instance_creation.activity_tracker_resource_key : ""
      url        = local.at_endpoint
    }

    provider "logdna" {
      alias      = "ld"
      servicekey = module.test_observability_instance_creation.logdna_resource_key != null ?     module.test_observability_instance_creation.logdna_resource_key : ""
      url        = local.at_endpoint
    }

---

### Checklist for reviewers

- [ ] The PR references a GitHub issue.
- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Merge by using "Squash and merge".
- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author.

    The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).

### CODE REVIEW

File: examples/observability_sysdig/providers.tf [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+
+provider "logdna" {
+  alias      = "at"
+  servicekey = module.test_observability_instance_creation.activity_tracker_resource_key != null ? module.test_observability_instance_creation.activity_tracker_resource_key : ""
+  url        = local.at_endpoint
+}
+
+provider "logdna" {
+  alias      = "ld"
+  servicekey = module.test_observability_instance_creation.logdna_resource_key != null ? module.test_observability_instance_creation.logdna_resource_key : ""
```
Comment (ocofaigh) [inline examples/observability_sysdig/providers.tf:13]: What happens when the logic passes a blank string for servicekey value (aka if a resource key is not created if consumer chooses not to provision logdna)? Does the module give any errors? I think the provider block is still required right? But not used?

File: examples/observability_logdna/version.tf [hunk 2]
```diff
@@ -6,5 +6,9 @@ terraform {
       source  = "ibm-cloud/ibm"
       version = "1.49.0"
     }
+    logdna = {
+      source  = "logdna/logdna"
+      version = "1.13.1"
```
Comment (ocofaigh) [inline examples/observability_logdna/version.tf:11]: will need to be locked into 1.14.2 after updating in version.tf

File: examples/observability_default/version.tf [hunk 3]
```diff
@@ -6,5 +6,9 @@ terraform {
       source  = "ibm-cloud/ibm"
       version = "1.49.0"
     }
+    logdna = {
+      source  = "logdna/logdna"
+      version = "1.13.1"
```
Comment (ocofaigh) [inline examples/observability_default/version.tf:11]: will need to be locked into 1.14.2 after updating in version.tf

File: examples/observability_at/version.tf [hunk 4]
```diff
@@ -6,5 +6,9 @@ terraform {
       source  = "ibm-cloud/ibm"
       version = "1.49.0"
     }
+    logdna = {
+      source  = "logdna/logdna"
+      version = "1.13.1"
```
Comment (ocofaigh) [inline examples/observability_at/version.tf:11]: will need to be locked into 1.14.2 after updating in version.tf

File: examples/observability_archive/version.tf [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  required_version = ">= 1.0.0"
+  required_providers {
+    # Pin to the lowest provider version of the range defined in the main module to ensure lowest version still works
+    ibm = {
+      source  = "ibm-cloud/ibm"
+      version = "1.49.0"
+    }
+    logdna = {
+      source  = "logdna/logdna"
+      version = "1.13.1"
```
Comment (ocofaigh) [inline examples/observability_archive/version.tf:11]: will need to be locked into 1.14.2 after updating in version.tf

File: examples/observability_archive/README.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# Provision LogDNA with a COS bucket for achive
```
Comment (ocofaigh) [inline examples/observability_archive/README.md:1]: achive -> archive

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Left some minor comments. Can we also run the new archive example in pr_test.go so it gets tested? Thanks

**Comment by ocofaigh** [comment]
I think we should also include the provider config in the usage block of the main readme now that it requires extra providers

File: examples/observability_at/version.tf (hunk 4 above)
Comment (shemau) [reply examples/observability_at/version.tf:11]: Fixed

File: variables.tf [hunk 7]
```diff
@@ -142,3 +142,52 @@ variable "activity_tracker_tags" {
...
+variable "logdna_cos_bucket_name" {
+  type        = string
+  description = "Only required to archive. The name of an existing COS bucket to be used for the LogDNA archive"
+  default     = null
+}
+
+variable "logdna_cos_bucket_endpoint" {
+  type        = string
+  description = "Only required to archive. The 'public' or 'private' endpoint for the COS bucket for the LogDNA archive"
+  default     = null
```
[the remark this answers is not in the document]
Comment (shemau) [reply variables.tf:174]: It is the public endpoint or the private endpoint that needs passing. Not the endpoint type.

Per the observability_archive example, where the public endpoint is passed from the COS module.

Does it read better as

  description = "Only required to archive. An endpoint for the COS bucket for the LogDNA archive. Pass either the public or private endpoint"

File: variables.tf [hunk 8]
```diff
@@ -142,3 +142,52 @@ variable "activity_tracker_tags" {
...
+variable "at_cos_bucket_name" {
+  type        = string
+  description = "Only required to archive. The name of an existing COS bucket to be used for the Activity Tracker archive"
+  default     = null
+}
+
+variable "at_cos_bucket_endpoint" {
+  type        = string
+  description = "Only required to archive. The 'public' or 'private' endpoint for the COS bucket for the Activity Tracker archive"
+  default     = null
```
[the remark this answers is not in the document]
Comment (shemau) [reply variables.tf:192]: It is the public endpoint or the private endpoint that needs passing. Not the endpoint type.

Per the observability_archive example, where the public endpoint is passed from the COS module.

Does it read better as:

  description = "Only required to archive. An endpoint for the COS bucket for the Activity Tracker archive. Pass either the public or private endpoint"

File: variables.tf [hunk 9]
```diff
@@ -142,3 +142,52 @@ variable "activity_tracker_tags" {
...
+# Archive options, access key and information about COS bucket
+variable "enable_archive" {
+  type        = bool
+  description = "Enable archive on logDNA and Activity Tracker instances"
+  default     = false
+}
+
+variable "ibmcloud_api_key" {
+  type        = string
+  description = "Only required to archive. The IBM Cloud API Token."
```
[the remark this answers is not in the document]
Comment (shemau) [reply variables.tf:154]: fixed

File: examples/observability_archive/README.md (hunk 6 above)
Comment (shemau) [reply examples/observability_archive/README.md:1]: fixed

File: examples/observability_archive/version.tf (hunk 5 above)
Comment (shemau) [reply examples/observability_archive/version.tf:11]: Updated

File: examples/observability_default/version.tf (hunk 3 above)
Comment (shemau) [reply examples/observability_default/version.tf:11]: Updated

File: examples/observability_logdna/version.tf (hunk 2 above)
Comment (shemau) [reply examples/observability_logdna/version.tf:11]: Updated

File: version.tf [hunk 10]
```diff
@@ -6,5 +6,10 @@ terraform {
       source  = "ibm-cloud/ibm"
       version = ">= 1.49.0"
     }
+    logdna = {
+      source                = "logdna/logdna"
+      version               = ">= 1.13.1"
```
[the remark this answers is not in the document]
Comment (shemau) [reply version.tf:11]: Updated

File: examples/observability_sysdig/providers.tf (hunk 1 above)
Comment (shemau) [reply examples/observability_sysdig/providers.tf:13]: There are no errors. The sysdig example (no logdna) plans and applies just fine. I assume the provider is evaluated just in time and only if the logdna_archive resource using it is in the dependancy tree.
I have checked manually that following a plan using archive that the archive option is set by viewing the logdna/at in the UI observability section and making sure archiving is enabled.
From memory, I think I tried null but that failed a pre-commit type check. So I made it a blank

File: variables.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (ocofaigh) [reply variables.tf:192]: ah sorry - rereading that I think its fine

File: variables.tf (hunk 7 above)
[the remark this answers is not in the document]
Comment (ocofaigh) [reply variables.tf:174]: rereading that, I think its fine

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**